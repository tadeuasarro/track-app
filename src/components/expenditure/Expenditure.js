/* eslint-disable camelcase */
import PropTypes from 'prop-types';
import './expenditure.css';

const expenses = [
  'Total',
  'Education',
  'Grocery',
  'Health',
  'Leisure',
  'Living',
  'Transport',
];

const Expenditure = ({ expenditure }) => {
  const {
    value,
    description,
    date,
    expense_id,
  } = expenditure;
  return (
    <div className="expenditure-container">
      <p className="expenditure">
        Value:
        &nbsp;
        {value}
      </p>
      <p className="expenditure">
        Date:
        &nbsp;
        {date}
      </p>
      <p className="expenditure">
        Description:
        &nbsp;
        {description}
      </p>
      <p className="expenditure">
        Category:
        &nbsp;
        {expenses[expense_id]}
      </p>
    </div>
  );
};

Expenditure.propTypes = {
  expenditure: PropTypes.shape({
    expense_id: PropTypes.number.isRequired,
    user_id: PropTypes.number.isRequired,
    value: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
  }),
};

Expenditure.defaultProps = {
  expenditure: {},
};

export default Expenditure;
