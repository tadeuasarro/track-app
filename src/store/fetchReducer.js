import { SET_LOADING, SET_ERRORS } from '../actions/fetch';

const initialState = {
  loading: false,
  errors: false,
};

const fetchReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_LOADING:
      return {
        loading: true,
        errors: false,
      };
    case SET_ERRORS:
      return {
        loading: false,
        errors: action.payload,
      };
    default:
      return state;
  }
};

export default fetchReducer;
