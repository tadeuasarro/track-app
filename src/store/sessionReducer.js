import { SET_CURRENT_USER, UPDATE_USER_TARGET } from '../actions/session';

const initialState = {
  username: false,
  id: false,
  target: false,
  token: false,
};

const sessionReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_CURRENT_USER:
      return {
        username: action.payload.user.username,
        id: action.payload.user.id,
        target: action.payload.user.target,
        token: action.payload.token,
      };
    case UPDATE_USER_TARGET:
      return {
        ...state,
        target: action.payload,
      };
    default:
      return state;
  }
};

export default sessionReducer;
