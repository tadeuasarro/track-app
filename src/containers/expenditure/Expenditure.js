/* eslint-disable camelcase */
import PropTypes from 'prop-types';
import { useState } from 'react';
import { useDispatch } from 'react-redux';
import Loading from '../../components/loading/Loading';
import deleteExpenditures from '../../api/deleteExpenditures';
import './expenditure.css';
import { setExpenditures } from '../../actions/expenditure';

const expenses = [
  'Total',
  'Education',
  'Grocery',
  'Health',
  'Leisure',
  'Living',
  'Transport',
];

const Expenditure = ({ expenditure }) => {
  const [state, setState] = useState({
    pending: false,
    error: false,
  });

  const dispatch = useDispatch();
  const {
    value,
    description,
    date,
    expense_id,
    id,
  } = expenditure;
  const handleClick = async id => {
    setState({
      ...state,
      pending: true,
    });

    const res = await deleteExpenditures(id);

    setState(res.state);

    if (!res.state.error) {
      dispatch(setExpenditures(res.payload));
    }
  };

  if (state.pending) return <Loading />;

  return (
    <div className="expenditure-container">
      <p className="expenditure">
        <span>Value:</span>
        &nbsp;
        {value}
      </p>
      <p className="expenditure">
        <span>Date:</span>
        &nbsp;
        {date}
      </p>
      <p className="expenditure">
        <span>Description:</span>
        &nbsp;
        {description}
      </p>
      <p className="expenditure">
        <span>Category:</span>
        &nbsp;
        {expenses[expense_id]}
      </p>
      <button type="button" onClick={() => handleClick(id)}>
        <i className="fas fa-trash-alt" />
      </button>
    </div>
  );
};

Expenditure.propTypes = {
  expenditure: PropTypes.shape({
    expense_id: PropTypes.number.isRequired,
    user_id: PropTypes.number.isRequired,
    value: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
  }),
};

Expenditure.defaultProps = {
  expenditure: {},
};

export default Expenditure;
