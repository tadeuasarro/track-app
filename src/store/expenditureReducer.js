import { SET_EXPENDITURES } from '../actions/expenditure';
import filterExpenditures from '../helpers/filterExpenditures';

const initialState = {
  expenditures: false,
  summary: false,
};

const expenditureReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_EXPENDITURES:
      const summary = filterExpenditures(action.payload);
      return {
        expenditures: action.payload,
        summary: summary,
      }
    default:
      return state;
  }
};

export default expenditureReducer;
