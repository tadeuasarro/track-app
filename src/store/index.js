import { combineReducers, createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import sessionReducer from './sessionReducer';
import expenditureReducer from './expenditureReducer';
import fetchReducer from './fetchReducer';

const store = createStore(
  combineReducers({
    session: sessionReducer,
    expenditure: expenditureReducer,
    fetch: fetchReducer,
  }),
  applyMiddleware(thunk),
);

export default store;
