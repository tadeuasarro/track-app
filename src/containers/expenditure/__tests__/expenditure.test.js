import { render, screen, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../../../store/index';
import Expenditure from '../Expenditure';

const expenditure = {
  value: 200,
  description: 'This is some description',
  date: '2020/03/01',
  category: 'Education',
  user_id: 1,
  id: 1,
};

describe('Rendering component', () => {
  it('creates an App component', () => {
    act(() => {
      render(
        <Provider store={store}>
          <Expenditure expenditure={expenditure} />
        </Provider>,
      );
    });
    // eslint-disable-next-line
    const element = screen.getByText('Description:');
  });
});

describe('Display', () => {
  it('renders correctly', () => {
    const comp = renderer.create(
      <Provider store={store}>
        <Expenditure expenditure={expenditure} />
      </Provider>,
    ).toJSON();
    expect(comp).toMatchSnapshot();
  });
});
