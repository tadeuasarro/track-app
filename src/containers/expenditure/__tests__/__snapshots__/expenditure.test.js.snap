// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display renders correctly 1`] = `
<div
  className="expenditure-container"
>
  <p
    className="expenditure"
  >
    <span>
      Value:
    </span>
     
    200
  </p>
  <p
    className="expenditure"
  >
    <span>
      Date:
    </span>
     
    2020/03/01
  </p>
  <p
    className="expenditure"
  >
    <span>
      Description:
    </span>
     
    This is some description
  </p>
  <p
    className="expenditure"
  >
    <span>
      Category:
    </span>
     
    Education
  </p>
  <button
    onClick={[Function]}
    type="button"
  >
    <i
      className="fas fa-trash-alt"
    />
  </button>
</div>
`;
